### ISSUE TITLE
Rewrite of sharing mechanism

### ISSUE THREAD

**Issue by paulduchesne** [paulduchesne/state#14]
Emulation of sharing process between two users

```python
import base64
import hashlib
import json
import pathlib
import rdflib
import uuid
from cryptography.fernet import Fernet

# emulation of state storage and subjective sharing.

paul_graph = rdflib.Graph()

# paul defines a statement, stores in local graph.

paul_graph.add((rdflib.URIRef('local://paul'), rdflib.URIRef('local://birthday'), rdflib.Literal('1986-04-14')))
print(paul_graph.serialize(format='ttl'))

# simulate write to public web instance.

for s,p,o in paul_graph.triples((None, None, None)):

    statement = rdflib.Graph().add((s, p, o)).serialize(format='nt')
    statement_id = hashlib.md5(statement.encode()).hexdigest()
    statement_path = pathlib.Path.cwd() / f'{statement_id}.ttl'

    if not statement_path.exists():

        keys_path = pathlib.Path.cwd() / 'keys.json'
        if not keys_path.exists():
            with open(keys_path, 'w') as empty_keys:
                json.dump([], empty_keys)

        with open(keys_path) as current_keys:
            current_keys = json.load(current_keys)

        key = str(uuid.uuid4()).replace('-', '')
        fernet = Fernet(base64.urlsafe_b64encode(key.encode()))
        encrypted_statement = fernet.encrypt(statement.encode()).decode()

        current_keys.append({'statement_id': statement_id, 'key':key})
        with open(pathlib.Path.cwd() / 'keys.json', 'w') as keys_out:
            json.dump(current_keys, keys_out, indent=4)

        public_state = rdflib.Graph().add((rdflib.URIRef(f'remote://{statement_id}'), rdflib.URIRef('remote://content'), rdflib.Literal(encrypted_statement)))
        public_state.serialize(destination=(pathlib.Path.cwd() / f'{statement_id}.ttl'), format='ttl') 
```

Second user picks up permitted statement, interleaves with own declaration.

```python
# data subjectively exposed to external parties.

lucy_graph = rdflib.Graph()

# certain keys shared with user.

with open(pathlib.Path.cwd() / 'keys.json') as keys_in:
    import_keys = json.load(keys_in)

for statement_import in import_keys:

    incoming = rdflib.Graph().parse(pathlib.Path.cwd() / f"{statement_import['statement_id']}.ttl")
    statement_share = [o for s,p,o in incoming.triples((None, None, None))]

    if len(statement_share) != 1:
        raise Exception('This should never happen.')

    statement_share = statement_share[0]

    fernet = Fernet(base64.urlsafe_b64encode(statement_import['key'].encode()))
    private_statement = fernet.decrypt(statement_share.encode()).decode()
    lucy_graph += rdflib.Graph().parse(data=private_statement)

    # external triple has been imported. Note that provenance would be very very useful here.

# second user adds their own data statement. 

lucy_graph.add((rdflib.URIRef('local://lucy'), rdflib.URIRef('local://birthday'), rdflib.Literal('1988-12-04')))
print(lucy_graph.serialize(format='ttl')) 
```

**Comment by paulduchesne** [comment]
Also stepping through a user defining other users, but incorporating their self-defined identities at a later date.

```sh
PAUL=9f863940-6bfc-44b6-ad90-b99410154689
LUCY=659a774e-a59c-44d5-8a54-d014831e19c0
```

Paul says his birthday is 

```rdf
9f863940-6bfc-44b6-ad90-b99410154689://9f863940-6bfc-44b6-ad90-b99410154689 hasbirthday '1986-04-14' NAMED GRAPH 9f863940-6bfc-44b6-ad90-b99410154689
```

Paul creates a person called Lucy, and defines their birthday.

```rdf
9f863940-6bfc-44b6-ad90-b99410154689://fe84a3d6-aa73-42bf-a8d1-0a25827b8904 hasbirthday '1988-12-04' NAMED GRAPH 9f863940-6bfc-44b6-ad90-b99410154689
```

Lucy creates a graph

```
659a774e-a59c-44d5-8a54-d014831e19c0://659a774e-a59c-44d5-8a54-d014831e19c0 has birthday '1988-12-04' NAMED GRAPH 659a774e-a59c-44d5-8a54-d014831e19c0
```

Both sides can now include each others data, but a key development would be Paul realizing Lucy has self-defined herself,
and updates the node in his graph to match this.

In short, we can define any information about any entity we want, but with options to import their own self-defined id, we connect entities, and possibly grant authority over statements they make about themselves (if they own the domain). For example, Lucy's sense of her own birthday should be more trustworthy than Paul's, which we can verify by the match on self-defined node and graph source.
